include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcStream.
+   */
+  virtual WasmResult grpcClose(uint32_t /* token */) { // cancel on call, close on stream.
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Cancel a gRPC stream or call.  No more calls will occur.
+   * @param token is a token returned from grpcSream or grpcCall.
+   */
+  virtual WasmResult grpcCancel(uint32_t /* token */) { // cancel on call, reset on stream.
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcSream.
+   * @param message is a (serialized) message to be sent.
+   * @param end_stream indicates that the stream is now end_of_stream (e.g. WriteLast() or
+   * WritesDone).
+   */
+  virtual WasmResult grpcSend(uint32_t /* token */, string_view /* message */,
+                              bool /* end_stream */) { // stream only
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Metrics
+
+  /**
+   * Define a metric (Stat).
+   * @param type is the type of metric (e.g. Counter).
+   * @param name is a string uniquely identifying the metric.
+   * @param metric_id_ptr is a location to store a token used for subsequent operations on the
+   * metric.
+   */
+  virtual WasmResult defineMetric(MetricType /* type */, string_view /* name */,
+                                  uint32_t * /* metric_id_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Increment a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param offset is the offset to apply to the Counter.
+   */
+  virtual WasmResult incrementMetric(uint32_t /* metric_id */, int64_t /* offset */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Record a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value is the value to store for a Gauge or increment for a histogram or Counter.
+   */
+  virtual WasmResult recordMetric(uint32_t /* metric_id */, uint64_t /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Get the current value of a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value_ptr is a location to store the value of the metric.
+   */
+  virtual WasmResult getMetric(uint32_t /* metric_id */, uint64_t * /* value_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Properties
+
+  /**
+   * Get the value of a property.  Some properties are proxy-independent (e.g. ["plugin_root_id"])
```

Comment:
How do properties related to metadata and Envoy's typed filter state? I would rather we didn't have yet another independent key:value bag scheme.